engine/runtime/src/main/java/org/enso/interpreter/node/expression/literal/IntegerLiteralNode.java

```diff
@@ -1,13 +1,18 @@
 package org.enso.interpreter.node.expression.literal;
 
+import com.oracle.truffle.api.Assumption;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.Truffle;
 import com.oracle.truffle.api.frame.VirtualFrame;
 import com.oracle.truffle.api.nodes.NodeInfo;
 import org.enso.interpreter.node.ExpressionNode;
 
 /** A representation of integer literals in Enso. */
 @NodeInfo(shortName = "IntegerLiteral")
 public final class IntegerLiteralNode extends ExpressionNode {
-  private final long value;
+  private static final Assumption CONSTANTS_ARE_CONSTANTS = Truffle.getRuntime().createAssumption("Constants were never updated");
+  @CompilerDirectives.CompilationFinal
```

Comment:
I'd consider having 2 fields here – one `final` the other not, and deciding based on the assumption. We're only wasting one long's worth of space per literal, which is tiny, but the resulting code is going to be much faster without the @TruffleBoundary'd call.

Reply:
> much faster without the @TruffleBoundary'd call.

Oh, yeah. Good catch. I haven't thought about that.

Having two fields feels however ugly. As [you wrote](https://github.com/enso-org/enso/pull/3508#discussion_r888799689) - we don't want to influence regular code, when not running in the IDE. The standard way to inject such _tooling specific_ behavior is to use instrumentation. Or a special node that replaces this one. Etc. Moreover we want to support this behavior on other types of the nodes, not just on `IntegerLiteralNode`.

Reply:
Returned back to original `final` field as of [31648ad](https://github.com/enso-org/enso/pull/3508/commits/31648adb39548983b077f4bb0d0fe2ae3fd5fb9a).